From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: CrimsonEdgeHope <92579614+crimsonedgehope@users.noreply.github.com>
Date: Thu, 20 Feb 2025 13:36:32 +0800
Subject: [PATCH] Fix sending head rotation packet to all viewers

This code, originally introduced by CraftBukkit, sends the entity's head
rotation when it gains a new viewer. But instead of sending the packet
only to the viewer, it broadcasts it to all viewers. This can cause many
packets to be sent when an entity spawns near many players.

This patch modifies the code to send the packet only to the viewer
that the packet was meant to be delivered to.

diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
index c1dc04045bf23dc0ec034781e13f65bbc5edb2fa..42596b1b7b0e596d185d35be33522d7829c77093 100644
--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
@@ -451,7 +451,7 @@ public class EntityTrackerEntry {
 
                     // CraftBukkit start - Fix for nonsensical head yaw
                     this.i = MathHelper.d((tracker.isFakeHead() ? tracker.fakeYaw : this.tracker.getHeadRotation()) * 256.0F / 360.0F);
-                    this.broadcast(new PacketPlayOutEntityHeadRotation(this.tracker, (byte) i));
+                    entityplayer.playerConnection.sendPacket(new PacketPlayOutEntityHeadRotation(this.tracker, (byte) this.i)); // PandaSpigot - Don't broadcast
                     // CraftBukkit end
 
                     if (this.tracker instanceof EntityLiving) {
